dsync: use SEEK_DATA for sparse detection if possible

ofaaland opened the pull request:
In the past, when flag --sparse is given, the fiemap ioctl is used to detect holes and data extents.

lseek() with SEEK_HOLE and SEEK_DATA works on Lustre 2.15.7 when fiemap does not (e.g. files striped via PFL), and works on tmpfs which appears not support fiemap, at least on RHEL 8.

Tools in wider use, such as GNU Fileutils and tar, prefer lseek() over fiemap.

Since MFU doesn't use the other information returned by fiemap anyway, and use SEEK_HOLE and SEEK_DATA first, and fall back on fiemap only if that fails.

ofaaland committed:
dsync: use SEEK_DATA for sparse file detection if possible

When flag --sparse is given, the fiemap ioctl has been used to detect
holes and data extents.

With this code, try detecting holes and data extents with lseek() and
SEEK_HOLE and SEEK_DATA first.  Fall back on using fiemap only if
lseek() fails.

This is consistent with more mainstream tools such as GNU Fileutils and
tar which prefer SEEK_HOLE, SEEK_DATA.  MPIFileUtils doesn't use the
additional information provided by fiemap anyway.

SEEK_HOLE, SEEK_DATA are also supported better by Lustre, e.g. for files
striped via PFL, and on tmpfs (as of RHEL 8.10).

Signed-off-by: Olaf Faaland <faaland1@llnl.gov>

daltonbohning commented on src/common/mfu_flist_copy.c:
```diff
@@ -1981,8 +1999,150 @@ static int mfu_copy_file_fiemap(
+    op = SEEK_DATA;
+    idx = 0;
+
+    MFU_LOG(MFU_LOG_INFO, "map from lseek():");
```
Is this meant to be removed?

ofaaland replied:
Oops, I meant for that to be MFU_LOG_DBG.  Thanks

ofaaland replied:
Fixed in the update I just pushed

ofaaland commented:
> I think the preference upstream (GNU Fileutils, tar, etc.) is to prefer SEEK_HOLE, SEEK_DATA over FIEMAP, unless MFU can use some of the extra information from the FIEMAP data vs. just knowing where the data is?

That's good to know, thank you.  At the moment, MFU doesn't use any additional information from FIEMAP.  I'll change the order.

defaziogiancarlo commented on src/common/mfu_flist_copy.c:
```diff
@@ -1981,8 +1999,150 @@ static int mfu_copy_file_fiemap(
+        offset = lseek(mfu_src_file->fd, offset, op);
+
+        if (op == SEEK_HOLE) {
+            MFU_LOG(MFU_LOG_DBG, "%llu - %llu type %s\n", start, offset, (op == SEEK_HOLE ? "data" : "hole"));
```
the ternary operator condition is always true here, "op == SEEK_HOLE" is always true inside this if block.

ofaaland replied:
I'll just remove that message.

ofaaland commented on src/common/mfu_flist_copy.c:
```diff
@@ -1888,7 +1888,37 @@ static int mfu_copy_file_normal(
+        return NULL;
+    }
+
+    memset((char *)extent_list + old_size, 0, extent_list_size - old_size);
```
Good catch.  I'll check for that.

ofaaland replied:
Fixed in the update I just pushed

ofaaland commented:
There are wrappers for lseek, such as `mfu_file_lseek()` which calls `daos_lseek()` for daos file systems.  I'm not using those, so I need to refactor.

Switching back to draft temporarily.

ofaaland commented:
Switched to use mfu_file_lseek() so the right variant of lseek is called for source files on DAOS, and so we get retries for certain conditions.

Fixed a bug which was triggered when a file was mostly sparse and multiple threads were used, which resulted in no data being copied.

I think it's now ready.

daltonbohning commented on src/common/mfu_flist_copy.c:
```diff
@@ -1981,8 +2007,184 @@ static int mfu_copy_file_fiemap(
+    uint64_t start;
+
+    /* Verify that SEEK_HOLE and SEEK_DATA are supported */
+    if ((mfu_file_lseek(src, mfu_src_file, offset, SEEK_HOLE) == (off_t)-1) ||
+        (mfu_file_lseek(src, mfu_src_file, offset, SEEK_DATA) == (off_t)-1)) {
```
Thanks for using these wrappers. FWIW, there is no support for this yet in daos

daltonbohning reviewed (approved):
I'm not familiar with these sparse functions but conceptually LGTM
